Subresource feature /dummy/1/relatedDummies

soyuka opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no (should not be any)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #634 - not really a fix but it's a start
| License       | MIT
| Doc PR        | /

#### Note:

~~I didn't implement the DataProvider yet! I'm just proposing those first changes to see if you have comments/improvements on the way I handled the nested behavior!~~

#### Related issues/PR:

#885
#634

#### What should this PR end up doing:

Allow `/entity/{id}/association` operations when a flag on a resource's property is set. For example:

```php

/**
 * @ApiResource
 */
class Dummy {
    /**
     * @ApiProperty(subcollection=true) 
    */
    public $relatedDummies; 
}
```

```
curl http://localhost/api/dummy/1/related_dummies
```

You can chain those and have `/api/dummy/1/related_dummies/2/third_levels`.

#### Changes:
- Add `subcollection` (boolean) on the PropertyMetadata
- Initialize a new class `ApiPlatform\Util\OperationTypes` (maybe the path isn't the best one, suggestions?) to ease the handling of operation types
- NestedCollectionDataProvider

#### TODO 
- [x] path customization through path resolvers (new method no breaks)
- [x] it will help refactoring ApiLoader kinda messy right now
- [x] query review
- [x] tests tests tests

soyuka commented on features/main/embedded.feature:
```diff
@@ -0,0 +1,108 @@
+  I need to be able to retrieve embedded resources only
+
+  @dropSchema
+  Scenario: Do nothing
```
Temporary hack to allow the single `./vendor/bin/behat --stop-on-failure -v --name="Embedded resources support"`

teohhanhui commented:
I don't think nestedOperation is a good name, but then again readableLink / writableLink are very confusing names as well...

teohhanhui commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -71,15 +78,26 @@ public function load($data, $type = null)
+                }
+
+                $operation = ['method' => 'GET', 'property' => $property];
+                $this->addRoute($routeCollection, $resourceClass, strtolower($property).'_get', $operation, $resourceShortName, OperationTypes::NESTED);
```
I can imagine the route name as such: `api_products_get_images_subcollection`

So `$operationName` should be 'get_'.strtolower($property)`.

teohhanhui replied:
I can imagine the route name as such: `api_products_get_images_subcollection`

So `$operationName` should probably be 'get_'.strtolower($property)`.

teohhanhui commented on src/Metadata/Property/PropertyMetadata.php:
```diff
@@ -31,8 +31,9 @@
     private $identifier;
     private $childInherited;
     private $attributes;
+    private $nestedOperation;
```
I don't think nestedOperation is a good name, but then again readableLink / writableLink are very confusing names as well...

Perhaps `$subcollection` / `hasSubcollection`?

teohhanhui commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -156,7 +173,8 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
                 '_api_resource_class' => $resourceClass,
-                sprintf('_api_%s_operation_name', $collection ? 'collection' : 'item') => $operationName,
+                sprintf('_api_%s_operation_name', $operationType) => $operationName,
+                '_api_resource_property' => $operation['property'] ?? null,
```
`_api_subcollection_property_name`?

soyuka commented:
Okay, now it does work, I still have to fix the tests but as I made many changes I'd like to know if those are okay / not breaking anything:

1. Moved `normalizeIdentifiers` to a new class `IdentifiersHelper` (`ItemDataProvider` constructor is impacted)
2. Changed the `ResourceClassResolver->getResourceClass()` signature as I needed more than only one class (may be a no-go)

 The rest is pretty straightforward:
- collection extensions should work on Subcollections
- for now subcollections with composite identifiers are not possible
- the query works fine for now, but I'd like to avoid a sub-select, I tried different options without success and I'll continue trying [ref for myself](http://stackoverflow.com/a/15444719/1145578)
- it's needed to have the subcollection resource class in the `PropertyMetadata` or we can't really get what it should be from the serializer/ApiLoader without Doctrine.

Please tell me what you think about the changes in general, is it too complex? Do you see things that can be simplified?

Thanks!

ping @api-platform/core-team

dunglas commented on src/Util/OperationTypes.php:
```diff
@@ -0,0 +1,20 @@
+ * file that was distributed with this source code.
+ */
+
+namespace ApiPlatform\Core\Util;
```
We should find a better name, maybe can we just move this class in `ApiPlatform\Core`.

soyuka commented:
> I think that there is a flaw in the general design. We should find a way to handle as many level as we need. Introducing a subcollection operation type looks not scalable enough. We should probably try to tweak the current item and collection operations to make them able to be nested.

Yes I kinda agree with you and you would like to end up with something like:
```
ItemOperation  CollectionOperation  ItemOperation etc.
/dummy/1       /relatedDummies        /1
```

I'll try to dig deeper into this but we definitely need a different DataProvider for exactly this kind of operation.  Just chaining Item/Collection would be really bad in performances.
Though, I think the new operation type is required because it'll behave differently from the current one in many points. 

I'll do a separated Pull request tomorrow to get rid of that IdentifierGenerator / IdentifierManager or whatever.

<details> 
  <summary>Thoughts for later </summary>

```
loop resource properties
if property has subcollection
  keep track of property + subcollection entity
  loop subcollection properties //recurse
fi

declare operation with array of collections
declare every subcollection independently? Meaning

/foo/1/bar/1/baz

would declare

/foo/{id}/bar
/foo/{foo_id}/bar/{bar_id}/baz
```
</details>

soyuka commented on src/DataProvider/SubcollectionDataProviderInterface.php:
```diff
@@ -0,0 +1,37 @@
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+interface SubcollectionDataProviderInterface
```
We might but the dataprovides should understand that we "need the collection belonging to an item", and that's very different from both of the current implementations IMO.

teohhanhui commented:
@dunglas:
> We should find a way to handle as many level as we need.

Let's take your example:
> /users/2/posts/1/comments

I think this is how it should be:
If `User::$posts` is marked as a subcollection and `Post::$comments` is marked as a subcollection, the above should automagically work.

soyuka commented:
Please don't mind the last commit and this comment I'm just keeping some notes.

todo:
- [ ] path customization through path resolvers (new method no breaks)
- [ ] it will help refactoring ApiLoader kinda messy right now
- [ ] query review
- [ ] tests tests tests

soyuka commented on src/Bridge/Doctrine/Orm/SubcollectionDataProvider.php:
```diff
@@ -0,0 +1,128 @@
+
+        $queryBuilder->where(
+            $queryBuilder->expr()->in('o', $previousQueryBuilder->getDQL())
+        );
```
IMO this is the best way to do what we want because we force the SQL execution plan to go through indexes on doctrine identifiers. The other solution would be to use joins.

This transforms to (pseudo-dql):

```
SELECT thirdLevel
WHERE thirdLevel IN (
  SELECT thirdLevel FROM relatedDummies WHERE relatedDummies = ? AND relatedDummies IN (
    SELECT relatedDummies FROM Dummy WHERE Dummy = ?
  )
 )
```

soyuka commented on src/Api/ResourceClassResolver.php:
```diff
@@ -63,6 +63,39 @@ public function getResourceClass($value, string $resourceClass = null, bool $str
     /**
      * {@inheritdoc}
      */
+    public function getResourceClassFromContext($value, array $context = null, bool $strict = false): string
```
Instead of BC break I chose to duplicate the method, I think we should deprecate the old one, let me know if you've a better idea.

soyuka replied:
In fact I don't need this, back to the original method.

soyuka commented:
@Simperfit any clue on how I broke the `DocumentationNormalizer`?

soyuka commented:
All green RFR ping @api-platform/core-team

Simperfit commented on features/main/embedded.feature:
```diff
@@ -0,0 +1,150 @@
+  I need to be able to retrieve embedded resources only
+
+  @dropSchema
+  Scenario: Just delete previous stuff pls
```
Do you really need that ?

soyuka replied:
Forgot to remove it it's easier when you want to launch just that feature

Simperfit commented on src/Bridge/Doctrine/Orm/SubcollectionDataProvider.php:
```diff
@@ -0,0 +1,140 @@
+    private $managerRegistry;
+    private $collectionExtensions;
+
+    /**
```
You may not need phpdoc in this case

Simperfit commented:
@soyuka I will take some time to review this and thanks, that's a great improvement IMO, could be very usefull ;).

Simperfit commented on src/Bridge/Doctrine/Orm/SubcollectionDataProvider.php:
```diff
@@ -0,0 +1,140 @@
+
+        $num = count($context['identifiers']);
+
+        while ($num--) {
```
It feels like a bit complicated, could it be split ?

soyuka replied:
Because of the while ? :smile: 

It is complicated because we're doing a recursive query where each previous part is included in the next one:

```
$previousQueryBuilder->andWhere($qb->expr()->in($previousAlias, $qb->getDQL()));
```

and 

```
$queryBuilder->where(
    $queryBuilder->expr()->in('o', $previousQueryBuilder->getDQL())
);
```

The rest is pretty straightforward, we first select the data we need, than we're adding identifiers to the where clause.

Simperfit replied:
👍

Simperfit commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -71,21 +79,94 @@ public function load($data, $type = null)
+                $rootShortname = $rootResourceMetadata->getShortName();
+                $resourceRouteName = Inflector::pluralize(Inflector::tableize($rootShortname));
+
+                $operation['identifiers'] = [['id', $rootResourceClass]];
```
```php
$operation['identifiers'][] = ['id', $rootResourceClass];
````

soyuka replied:
At this point `$operation['identifiers']` is not declared.

Simperfit commented on src/DataProvider/ChainSubcollectionDataProvider.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Tries each configured data provider and returns the result of the first able to handle the resource class.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
```
Not you :p ?

soyuka commented:
Not really, look at the first post in this pull request, the feature is easy to use:

```php

/**
 * @ApiResource
 */
class Dummy {
    /**
     * @ApiProperty(subcollection="RelatedDummy") 
    */
    public $relatedDummies; 
}
```

```
curl http://localhost/api/dummy/1/related_dummies
```

teohhanhui commented:
@soyuka
> subcollection="RelatedDummy"

Why? Isn't the collection value type already available from the `PropertyMetadata`?

soyuka commented:
Wow your're right. I should also check that isCollection is true. Let me change this, it'll simplify things. Thanks.

teohhanhui commented:
Do filters work on the subcollection? :smile:

soyuka commented:
I've added the collection extensions, it should work out of the box. I'll add a test though. 

For now only the `get` operation is supported.

soyuka commented:
@teohhanhui filters did work, but now it's tested.

Note:

On subcollections, filters are the one from the retrieve sub collection, not those from the original entity. For example on `/dummies/1/relatedDummies`, the filters must be declared on the `RelatedDummies` resource.

teohhanhui commented:
Yes. I think that makes sense. So the effect is only as if there's a hidden
filter which restricts results based on the relation on the parent resource.

On Mon, 13 Feb 2017, 21:06 Antoine Bluchet, <notifications@github.com>
wrote:

> @teohhanhui <https://github.com/teohhanhui> filters did work, but now
> it's tested.
>
> Note:
>
> On subcollections, filters are the one from the retrieve sub collection,
> not those from the original entity. For example on
> /dummies/1/relatedDummies, the filters must be declared on the
> RelatedDummies resource.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/api-platform/core/pull/904#issuecomment-279387399>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAhf6zScXM56PGmc6O097A6aMyghPt6Gks5rcFVMgaJpZM4LfkZC>
> .
>

soyuka commented:
@dunglas you know what's wrong with appveyor? :| Seems it can't install php anymore

teohhanhui commented:
Is it possible to disable the generic collection `GET` operation (`/relatedDummies`) while still having `/dummy/1/relatedDummies` work?

soyuka commented:
@teohhanhui I don't see why it would not be possible. The `computeSubCollectionOperations` is done independently from other operations in `ApiLoader`.

soyuka commented on src/Util/OperationTypes.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace ApiPlatform\Core\Util;
+
+final class OperationTypes
```
@teohhanhui WDYT about the name? Do you agree we should put this in `ApiPlatform\Core`?

teohhanhui replied:
`ApiPlatform\Core\OperationType`, like https://github.com/giggsey/libphonenumber-for-php/blob/master/src/PhoneNumberType.php

soyuka replied:
`ApiPlatform\Core\Api\OperationType` looks nice to me! @dunglas do you approve?

dunglas replied:
Yep!

dunglas commented on src/EventListener/ReadListener.php:
```diff
@@ -30,10 +32,11 @@
     private $itemDataProvider;
 
-    public function __construct(CollectionDataProviderInterface $collectionDataProvider, ItemDataProviderInterface $itemDataProvider)
+    public function __construct(CollectionDataProviderInterface $collectionDataProvider, ItemDataProviderInterface $itemDataProvider, SubcollectionDataProviderInterface $subcollectionDataProvider)
```
Should be optional for BC.

dunglas replied:
For all new arguments, maybe can you add deprecation notice if new deps are not passed so we'll be able to remove this code in API Platform 3?

soyuka replied:
Sure, is there a particular way I should know to add deprecation notice?

dunglas replied:
Yep: http://symfony.com/doc/current/components/phpunit_bridge.html

dunglas commented on src/EventListener/ReadListener.php:
```diff
@@ -56,10 +59,12 @@ public function onKernelRequest(GetResponseEvent $event)
-        } else {
-            $data = $this->getItemData($request, $attributes);
+        foreach (OperationTypes::TYPES as $operationType) {
+            if (isset($attributes["{$operationType}_operation_name"])) {
```
Brackets can be removed.

teohhanhui replied:
Brackets make it easier to read and is (of course very slightly) faster to boot.

teohhanhui replied:
Actually, it can be very dangerous without brackets: http://php.net/manual/en/language.types.string.php#language.types.string.parsing

dunglas commented on src/EventListener/ReadListener.php:
```diff
@@ -56,10 +59,12 @@ public function onKernelRequest(GetResponseEvent $event)
-            $data = $this->getItemData($request, $attributes);
+        foreach (OperationTypes::TYPES as $operationType) {
+            if (isset($attributes["{$operationType}_operation_name"])) {
+                $method = sprintf('get%sData', ucfirst($operationType));
```
If would prefer a switch or something like that instead of dynamic call (it's hard to debug and it prevents PHP to optimize at compile time).

dunglas commented on src/Util/OperationTypes.php:
```diff
@@ -0,0 +1,20 @@
+
+final class OperationTypes
+{
+    const ITEM = 'item';
```
Why not using int as values (less memory usage)? Do we use those const in YAML?

soyuka replied:
We might do this with int values indeed. No need for those in YAML.

teohhanhui replied:
Is it really necessary to optimize this? Is it a tangible difference?

dunglas replied:
Tangible I don't think so. But it's common convention to use ints for constants no? Probably my C background...

soyuka replied:
Strings are making making things easier too do:

```
foreach (OperationTypes::TYPES as $operationType) {
+            $attribute = "_api_{$operationType}_operation_name";
```

I think it's fine like this but if we want integers we need to change this now or it might be painful later \o/.

dunglas replied:
Ok keep the string.

dunglas commented on src/Util/RequestAttributesExtractor.php:
```diff
@@ -39,21 +39,32 @@ private function __construct()
-        $itemOperationName = $request->attributes->get('_api_item_operation_name');
+        $hasRequestAttributeKey = false;
+        foreach (OperationTypes::TYPES as $operationType) {
+            $attribute = "_api_{$operationType}_operation_name";
```
Brackets can be removed.

dunglas commented on src/Util/RequestAttributesExtractor.php:
```diff
@@ -39,21 +39,32 @@ private function __construct()
+        foreach (OperationTypes::TYPES as $operationType) {
+            $attribute = "_api_{$operationType}_operation_name";
+            if ($request->attributes->has($attribute)) {
+                $result["{$operationType}_operation_name"] = $request->attributes->get($attribute);
```
Brackets can be removed.

dunglas commented on src/Util/RequestAttributesExtractor.php:
```diff
@@ -39,21 +39,32 @@ private function __construct()
-            throw new RuntimeException('One of the request attribute "_api_collection_operation_name" or "_api_item_operation_name" must be defined.');
+        if (false === $hasRequestAttributeKey) {
+            throw new RuntimeException('One of the following request attribute must be defined: '.implode(', ', array_map(function ($operationType) {
+                return "_api_{$operationType}_operation_name";
```
Brackets can be removed.

soyuka commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -34,19 +37,24 @@
 {
     const ROUTE_NAME_PREFIX = 'api_';
     const DEFAULT_ACTION_PATTERN = 'api_platform.action.';
+    const SUBCOLLECTION_SUFFIX = '_get_subcollection';
```
Indeed I did that on purpose because we only support `get` for now. It's better hardcoded here then below IMO.

soyuka commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -71,21 +79,96 @@ public function load($data, $type = null)
+            if (null === $parentOperation) {
+                $rootResourceMetadata = $this->resourceMetadataFactory->create($rootResourceClass);
+                $rootShortname = $rootResourceMetadata->getShortName();
+                $resourceRouteName = Inflector::pluralize(Inflector::tableize($rootShortname));
```
There is no strategy for that, I did the same thing that's already done in `addRoute` where we call the inflector to find the name:

```
        $resourceRouteName = Inflector::pluralize(Inflector::tableize($resourceShortName));
```

dunglas replied:
Sorry about that, I was sure it was configurable :p. Maybe can you group all those calls in a private method to be able to make the strategy configurable later if we want to.

soyuka replied:
This I can do!

soyuka commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -71,21 +79,96 @@ public function load($data, $type = null)
+
+            $subcollection = $propertyMetadata->getType()->isCollection() ? $propertyMetadata->getType()->getCollectionValueType()->getClassName() : $propertyMetadata->getType()->getClassName();
+
+            $propertyName = Inflector::pluralize(Inflector::tableize($property));
```
Was was the strategy you had in mind?

soyuka commented:
Fixed the comments except for the brackets part. May you check that the deprecation warnings are correct? Thanks!

dunglas commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -34,22 +37,31 @@
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+
+        if (null === $this->propertyNameCollectionFactory || null === $this->propertyMetadataFactory) {
+            @trigger_error('Missing dependencies to compute subcollection operations.', E_USER_DEPRECATED);
```
I would say `Not injecting "PropertyNameCollectionFactoryInterface" and "PropertyMetadataFactoryInterface" instances in __CLASS__ is deprecrated since API Platform 2.3 and will not be possible anymore in API Platform 3`.

soyuka replied:
Okay thanks!

dunglas commented on src/EventListener/ReadListener.php:
```diff
@@ -29,11 +30,17 @@
+        $this->subcollectionDataProvider = $subcollectionDataProvider;
+
+        if (null === $subcollectionDataProvider) {
+            @trigger_error('No subcollection data provider given', E_USER_DEPRECATED);
```
Same here.

soyuka commented on src/EventListener/ReadListener.php:
```diff
@@ -103,4 +112,30 @@ private function getItemData(Request $request, array $attributes)
+    {
+        if (null === $this->subcollectionDataProvider) {
+            throw new RuntimeException('No subcollection data provider.');
+        }
```
@dunglas should I throw this here? 

In ApiLoader I just fail silently but here I wasn't sure.

lbarulski commented:
@soyuka will it support one-to-one relation?

`/opinion/{opinion.id}/answer` - one answer for one opinion

soyuka commented:
@lbarulski they should yes. It's fairly rare to have one to one relations though, if you've an example I could add another test.

lbarulski commented:
https://gist.github.com/lbarulski/8cb2c3034a4fcae31f7920277efa2c30

soyuka commented:
Okay so OneToOne relations do work. But:

1. I had to inverse the relation you gave me to have `/questions/1/answers`. Indeed, we can not identify the `question.answer` if it's not this side that has the identity column. 
2. Because I had sub**collections** in mind, the returning result will be a collection. Though on a OneToOne relation we know that we expect only one result. Therefore it should be an Item. This is wrong IMO in terms of metadata.
3. The url, as shown above has `answers` plural, which is again wrong for this particular use case.

I didn't wanted to support a kinda `SubItem` thing because I don't think it's worth it. With filters you could have almost the same result through the origin item, here `Question` really easily.

So, what should we do about this? 
- Option 1: Prevent subcollections on OneToOne relations (ie throw an exception).
- Option 2: Support subitem retrieval to have a clean url and proper Item metadata (eg hydra item vs collection) + Option 1
- Option 3 : Do nothing, OneToOne works (only one way and throws a weird error that will make beginners panic the other way) but returns a collection.

I'm not publishing the behat test yet but it's done thanks for the example classes @lbarulski .

dunglas commented:
1-1 relations can already be handled thru embedding. I'll be for explicitly excluding them (throw).

teohhanhui commented:
Don't throw. The `subcollection` attribute shall have no effect when set on a to-one relation. :smile:

soyuka commented:
@teohhanhui So you'd say that they should not be allowed when creating routes in `ApiLoader` for example? How can I detect the `OneToOne` relation? I wanted to use `PropertyMetadata->isCollection()`, but if I do so I also need to do something about `ManyToOne` relations.

soyuka commented:
Hmm, as this was really easy I made some changes in the last commit that takes into consideration `PropertyMetadata->isCollection()` on embedded resources:
- don't pluralize route/path
- returns a hydra item when it's not a collection

I might revert this last commit but as I think it's interesting to have sub collections on ManyToOne relations, we better support it correctly by returning an item. Then, OneToOne support is automatic.

teohhanhui commented:
Wonderful! We may want to rename it to `subresource=true` then?

Simperfit commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -34,22 +37,31 @@
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+
+        if (null === $this->propertyNameCollectionFactory || null === $this->propertyMetadataFactory) {
+            @trigger_error('Not injecting "PropertyNameCollectionFactoryInterface" and "PropertyMetadataFactoryInterface" instances in '.__CLASS__.' is deprecrated since API Platform 2.3 and will not be possible anymore in API Platform 3', E_USER_DEPRECATED);
```
`deprecrated since API Platform 2.1 and will not be possible anymore in API Platform 3`

soyuka commented:
> Wonderful! We may want to rename it to subresource=true then? Or better, subresource_operation=true.

Haha yes I think we can rename to `subresource=true` as it's no longer a collection-only :D. 

All agree?

teohhanhui commented:
I think `subresource_operation` is more descriptive of what it actually is.

teohhanhui commented:
Okay, nevermind. `subresource=true` tells us that this relation is a subresource. It leaves more room for additional functionality surrounding a subresource, as it may not just be operations.

soyuka commented:
Rebased on master, fixed some scrutinizer issues.

@teohhanhui may you take a closer look to my work here to see if you've improvements/ideas/comments?

Thanks!

Simperfit commented:
I think that -1% of code coverage is a bit high :(

Simperfit commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -71,21 +83,117 @@ public function load($data, $type = null)
+                $operation['route_name'] = sprintf('%s%s_%s%s', self::ROUTE_NAME_PREFIX, $resourceRouteName, $propertyName, self::SUBRESOURCE_SUFFIX);
+
+                $operation['path'] = $this->operationPathResolver->resolveOperationPath($rootShortname, $operation, OperationType::SUBRESOURCE);
+            } else {
```
can't that else be remove and the value in it be the default ?

soyuka replied:
Not really because it's then based on the parentOperation, I might be able to refactor this but I think it's readable like this.

Simperfit commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -71,21 +83,117 @@ public function load($data, $type = null)
+
+                $operation['path'] = $this->operationPathResolver->resolveOperationPath($rootShortname, $operation, OperationType::SUBRESOURCE);
+            } else {
+                $identifier = $parentOperation['property'];
```
seems like it is used just one  time, maybe you shouldn't be creating a variable ?

Simperfit commented on src/Swagger/Serializer/DocumentationNormalizer.php:
```diff
@@ -94,19 +95,19 @@ public function normalize($object, $format = null, array $context = [])
-            $path = $this->getPath($resourceShortName, $operation, $collection);
-            $method = $collection ? $this->operationMethodResolver->getCollectionOperationMethod($resourceClass, $operationName) : $this->operationMethodResolver->getItemOperationMethod($resourceClass, $operationName);
+            $path = $this->getPath($resourceShortName, $operation, $operationType);
+            $method = $operationType === OperationType::ITEM ? $this->operationMethodResolver->getItemOperationMethod($resourceClass, $operationName) : $this->operationMethodResolver->getCollectionOperationMethod($resourceClass, $operationName);
```
why this change on the condition :o ?

soyuka replied:
Collection and Subresource are both using `CollectionOperation`, I just reversed it behavior hasn't changed.

soyuka commented:
About coverage, code here should have been covered, though lines like `@trigger_error` are useless to cover. Also, when merging master, coverage decreased from 0.6% more... I can add tests for things that are not related to this but it's messy.

soyuka commented:
@lbarulski 
@samusenkoiv

It would be lovely if you could try this feature and tell me if it fits your needs or if you've issues/improvements to add :).

soyuka commented:
@dunglas Constructor argument have been covered for BC breaks in this pull request. However, as I named  `$operationType` (and added constants etc.), I've changed a lot of function that were using `isCollection` with a boolean. Now it's a `string` representing the type being used.

dunglas commented on features/main/embedded.feature:
```diff
@@ -0,0 +1,280 @@
+    """
+    {
+      "@context": "/contexts/RelatedDummy",
+      "@id": "/related_dummies",
```
IIUC, the `@id` key should be `/dummies/1/related_dummies`

soyuka replied:
Hmm not sure because this is still a `related_dummy`. Anyway I really don't know how I can implement this because `getIriFromItem` isn't aware that we have requested this resource from a parent resource.

soyuka replied:
Oh wait this is set in hydra, my bad. Yeah I just need to add some data to the `SerializerContextBuilder` :).

dunglas commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -0,0 +1,164 @@
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+class SubresourceDataProvider implements SubresourceDataProviderInterface
```
Can you made this class `final`?

dunglas commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -0,0 +1,164 @@
+            $alias = $queryNameGenerator->generateJoinAlias($identifier);
+
+            //MANY_TO_MANY relations needs an explicit join so that the identifier part can be retrieved
+            if ($classMetadata->getAssociationMapping($previousIdentifier)['type'] === ClassMetadataInfo::MANY_TO_MANY) {
```
Can you invert this test (yoda style)?

dunglas commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -34,22 +37,31 @@
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+
+        if (null === $this->propertyNameCollectionFactory || null === $this->propertyMetadataFactory) {
+            @trigger_error('Not injecting "PropertyNameCollectionFactoryInterface" and "PropertyMetadataFactoryInterface" instances in '.__CLASS__.' is deprecrated since API Platform 2.1 and will not be possible anymore in API Platform 3', E_USER_DEPRECATED);
```
Is this really necessary? To ease using API Platform as a standalone lib, it would be nice to be able to not provide those instance (and then loose subresource support)

dunglas commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -71,21 +83,113 @@ public function load($data, $type = null)
+     * @param string          $rootResourceClass null on the first iteration, it then keeps track of the origin resource class
+     * @param string          $rootResourceClass null on the first iteration, it then keeps track of the parent operation
+     */
+    private function computeSubresourceOperations(RouteCollection $routeCollection, string $resourceClass, $rootResourceClass = null, $parentOperation = null)
```
Missing typehints for `$rootResourceClass` and `$parentOperation`

dunglas commented on src/Bridge/Symfony/Routing/ApiLoader.php:
```diff
@@ -71,21 +83,113 @@ public function load($data, $type = null)
+     * @param RouteCollection $routeCollection
+     * @param string          $resourceClass
+     * @param string          $rootResourceClass null on the first iteration, it then keeps track of the origin resource class
+     * @param string          $rootResourceClass null on the first iteration, it then keeps track of the parent operation
```
Duplicated line, and missing doc for `$parentOperation`

dunglas commented on src/EventListener/ReadListener.php:
```diff
@@ -28,11 +30,17 @@
         $this->itemDataProvider = $itemDataProvider;
+        $this->subresourceDataProvider = $subresourceDataProvider;
+
+        if (null === $subresourceDataProvider) {
```
Same here, I would just disable this feature if the `subresourceDataProvider` is not passed, without triggering a deprecation notice.

soyuka replied:
Okay, should I still throw a `RuntimeException` then?

dunglas replied:
Where?

soyuka replied:
https://github.com/soyuka/core/blob/52c0da20861400266d6d6cf137a27cda6bbee386/src/EventListener/ReadListener.php#L126

dunglas replied:
You can keep this exception. It is only triggered when a subresource is encountered, sounds legit to me.

dunglas commented on src/PathResolver/CustomOperationPathResolver.php:
```diff
@@ -28,12 +28,12 @@ public function __construct(OperationPathResolverInterface $deferred)
      * {@inheritdoc}
      */
-    public function resolveOperationPath(string $resourceShortName, array $operation, bool $collection): string
+    public function resolveOperationPath(string $resourceShortName, array $operation, string $operationType): string
```
It's a BC break but it can be easily fixed: remove the typehint and support both `string` and `bool`.

Here I would deprecate to pass a `bool`.

dunglas commented on src/PathResolver/DashOperationPathResolver.php:
```diff
@@ -23,16 +24,31 @@
      * {@inheritdoc}
      */
-    public function resolveOperationPath(string $resourceShortName, array $operation, bool $collection): string
+    public function resolveOperationPath(string $resourceShortName, array $operation, string $operationType): string
```
Same here, BC break that can fixed with the same trick.

dunglas commented on src/PathResolver/OperationPathResolverInterface.php:
```diff
@@ -21,11 +21,11 @@
+     * @param string $resourceShortName When the operation type is a subresource and the operation has more than one identifier, this value is the previous operation path
      * @param array  $operation         The operation metadata
-     * @param bool   $collection
+     * @param bool   $operationType     One of the constants defined in ApiPlatform\Core\Api\OperationType
```
Bad documented type.

dunglas commented on src/PathResolver/OperationPathResolverInterface.php:
```diff
@@ -21,11 +21,11 @@
      * @return string
      */
-    public function resolveOperationPath(string $resourceShortName, array $operation, bool $collection): string;
+    public function resolveOperationPath(string $resourceShortName, array $operation, string $operationType): string;
```
BC break, can be fixed the same way too.

dunglas commented on src/PathResolver/UnderscoreOperationPathResolver.php:
```diff
@@ -23,14 +24,24 @@
      * {@inheritdoc}
      */
-    public function resolveOperationPath(string $resourceShortName, array $operation, bool $collection): string
+    public function resolveOperationPath(string $resourceShortName, array $operation, string $operationType): string
```
Same here :)

dunglas commented on src/Serializer/SerializerContextBuilder.php:
```diff
@@ -42,9 +42,11 @@ public function createFromRequest(Request $request, bool $normalization, array $
-        if (isset($attributes['collection_operation_name'])) {
-            $context = $resourceMetadata->getCollectionOperationAttribute($attributes['collection_operation_name'], $key, [], true);
-            $context['collection_operation_name'] = $attributes['collection_operation_name'];
+        if (isset($attributes['collection_operation_name']) || isset($attributes['subresource_operation_name'])) {
```
WDYT of avoiding an extra call to `isset`?

```php
$operationKey = null;
if (isset($attributes['collection_operation_name'])) {
    $operationKey = 'collection_operation_name' ;
} elseif (isset($attributes['subresource_operation_name'])) {
    $operationKey = 'subresource_operation_name';
}

if (null !== $operationKey) // ...
```

soyuka commented:
@dunglas May you check the last commit (about corrected Hydra `@id`)? Let me know what you think about the implementation, if it looks fine to you I need to add some tests to the IriConverter :).

dunglas commented on src/PathResolver/DashOperationPathResolver.php:
```diff
@@ -23,16 +25,33 @@
         return $path;
     }
+
+    private function dashize(string $string)
```
`: string`

soyuka commented on src/Api/OperationTypeDeprecationHelper.php:
```diff
@@ -0,0 +1,42 @@
+ * - OperationType::COLLECTION
+ * - OperationType::SUBCOLLECTION
+ */
+final class OperationTypeDeprecationHelper
```
Not really useful to have a trait for this, static methods are fine enough IMO but I can change this.

backbone87 commented:
Is the statement from earlier in this ticket still true, that composite PKs do not work? If so, what needs to be done to support them? I would really like that feature for my next project and can offer some manpower to it. though, if it is too complicated, i could redesign my use-case to use an artificial single column PK.

Is it possible to define different operation/configuration depending on the "scope" a resource is used in?
Example: 2 related top-level entities (in UML terms: association)
- `/books/{book_id}/authors/{author_id}` only allow collection get and item get
- `/authors/{author_id}` allow all default operations
Or is that another use case not covered by this PR (and only covers the UML constructs "aggregation" and "composition")?

What about self-referencing associations? From my use-case:
- Top-level (aggregate root) entity `container`: contains mostly meta-data for authorization and concurrency-control and privately owns (UML composition) a set of nodes that form a DAG
- Subentity `node`: Insert-only data-point; has an id that is local to the container and is "auto-incremented" (not rly, it is just the version of the container at insertation time); has a type (handled outside of doctrine, no inheritance mapping; strategy pattern); has a data column (arbitrary json data, structure depends on the type); has some more domain-specific meta-data; **and of course** a (possibly empty) to-many association "derivedFrom" to other nodes of the same container (this association reflects the DAG).
- Some example filter queries:
  - `/containers/1/nodes?subgraph-of-most-recent-node-of-type=result` (this should also be the embedded node set, when getting a container `/container/1`)
  - `/containers/1/nodes?subgraph-of-nodes-derived-from=<NODE_ID>`
  - `/containers/1/nodes?subgraph-of-nodes-used-to-derive-node=<NODE_ID>`

soyuka commented:
> Is the statement from earlier in this ticket still true, that composite PKs do not work?

I think they do. 

> Is it possible to define different operation/configuration depending on the "scope" a resource is used in?

A subresource has only GET operations. No plan on implementing PUT/POST yet.

> What about self-referencing associations?

Just try the branch :) seams it should work well with your use-cases.

backbone87 commented:
> Just try the branch :) seams it should work well with your use-cases.

I am doin that right now ;)

> A subresource has only GET operations. No plan on implementing PUT/POST yet.

So i cant POST to `/container/{container_id}/nodes` to add a new node? (i dont need update, just create)

backbone87 commented:
There is no `/nodes` in my current approach, nodes are privately owned by a container entity. a node has composite PK consisting of the containers UUID and an index

soyuka commented:
Hmm, you might be able to get what you want by manually declaring a new operation for posts on your `/container/{uuid}/nodes`. Though this is a bit out of scope as `nodes` should be a `Resource`, so that it can be a `SubResource` of `containers`:

```php
/**
 * @Api/Resource
 */
class Nodes
{
}

/**
 * @ApiResource
 */
class Container
{
    /**
     * @ApiProperty(subresource=true)
     */
    private $nodes;
}
```

backbone87 commented:
Either i am completely missunderstanding this feature, or there are some major flaws in it.

Consider the following classes:
```php
/**
 * @API\ApiResource
 * @ORM\Entity
 */
class Container
{

    /**
     * @ORM\Id
     * @ORM\Column(name="id", type="guid")
     *
     * @var string UUID
     */
    private $id;

    /**
     * @API\ApiProperty(subresource=true)
     * @ORM\OneToMany(
     *      targetEntity="Node",
     *      mappedBy="container",
     *      indexBy="serial",
     *      fetch="LAZY",
     *      cascade={},
     *      orphanRemoval=false
     * )
     * @ORM\OrderBy({"serial"="ASC"})
     *
     * @var Collection|Datum[]
     */
    private $nodes;

    /**
     * @return string
     */
    public function getId(): ?string
    {
        return $this->id;
    }

    /**
     * @return array|Node[]
     */
    public function getNodes(): array
    {
        return $this->nodes->toArray();
    }
}

/**
 * @API\ApiResource
 * @ORM\Entity
 */
class Node
{
    /**
     * @ORM\Id
     * @ORM\ManyToOne(targetEntity="Container", fetch="LAZY")
     * @ORM\JoinColumn(name="container_id", referencedColumnName="id", onDelete="RESTRICT")
     *
     * @var Container
     */
    private $container;

    /**
     * @ORM\Id
     * @ORM\Column(name="serial", type="integer")
     *
     * @var integer Node serial
     */
    private $serial;

    /**
     */
    public function __construct()
    {
    }

    public function setContainer(Container $container): void
    {
        $this->container= $container;
        $this->serial = $container->nextSerial();
    }

    /**
     *
     * @return integer
     */
    public function getSerial(): int
    {
        return $this->serial;
    }
}
```

invoking `/containers/SOME_UUID/nodes` fails because the generated query looks something like this:
`SELECT o FROM Node o WHERE o IN(SELECT IDENTITY(id_a1.nodes) FROM Container id_a1 WHERE id_a1.id = :id_p1)`

but i would expect a query more like this:
`SELECT o FROM Node o WHERE o.container IN (SELECT id_a1.id FROM Container id_a1 WHERE id_a1.id = :id_p1)`

any hints?

soyuka commented:
@backbone87 thanks for your interest in this pull request. I'm currently reproducing your case in an intergration test. It's possible that there are uncovered parts here, because relations between resources can differ a lot according to the use cases. 

> Either i am completely missunderstanding this feature, or there are some major flaws in it.

Your choice of words is a bit strong here ^^. I don't think you misunderstood the feature:

`/containers/SOME_UUID/nodes` should get the nodes belonging to the container identified by `SOME_UUID`. 

To me the following DQL looks correct:

```sql
SELECT o FROM Node o -- select nodes
WHERE o IN( -- where nodes are
  SELECT IDENTITY(id_a1.nodes) -- the nodes
  FROM Container id_a1
  WHERE id_a1.id = :id_p1  -- belonging to this container 
)
```

The second one is also valid, but it assumes that there is a two-way relationship (ie container is stored in node). 

Anyway, I'm investigating this!

soyuka commented:
Thanks @backbone87 this was a really interesting case it had:
- a `OneToMany` relation which behavior is slightly different
- a composite key

I've fixed your issue and it works great through behat. Could you give it a try? Thanks!

backbone87 commented:
After digging deeper into the code, idk if this PR uses the right approach.

A configuration scheme i would like to see:
```php
/**
 * @ApiProperty(
 *   resource=@ApiResource(
 *     collectionValued=true|false,
 *     collectionOperations={ ... }, // invalid when collectionValued = false
 *     itemOperations={ ... }
 *   )
 * )
 */
private $nodes;
```

soyuka commented:
It's not really necessary to have to specify a subresource `Resource` because we already know the metadata (if it's a collection or not). Also, we would like to keep things simple for the end user.

What bother's you with the current implementation?

backbone87 commented:
Your changes do work with this use case.

Anyway, i think we should aim at getting full operation support (incl custom operations) on subresources and in combination with that: disable a resource being top lvl (or in terms of the given use case: remove /nodes)

backbone87 commented:
our communication is somewhat deferred ;) i am on gitter, if you want to have chat more directly

another problem with current configuration scheme: i cant name the paths properly (its autogenerated for subresources). Since i have to use german resources names for my current project the english inflector and pluralizer generates completely unuseable routes.

soyuka commented:
ping @api-platform/core-team can someone review phpstan fixes in the last commit? Thanks!

dunglas commented on src/Api/OperationTypeDeprecationHelper.php:
```diff
@@ -0,0 +1,45 @@
+     */
+    public static function getOperationType($operationType): string
+    {
+        if (is_bool($operationType)) {
```
To prevent the PHPStan error you may use `if( !is_string(($operationType)) {`.

soyuka replied:
hmm okay though it should be correct like this! I'll change it thx.

soyuka replied:
Okay so I did try `!is_string` but I get the same error :| I'll just ignore this one for now as this is a temporary class anyway.

dunglas commented:
Don't forget the doc PR :)

cocciagialla commented:
@dunglas I am using the annotation `@ApiProperty(subresource=true)` on a collection inside a Resource. It works well and I can call `/myresource/<id>/mysubresourcecollection`.
The problem is that in the swagger documentation this new endpoint is not exposed. Do you think it will be added in the future or there is a way to actually expose it? Thanks

soyuka commented:
Hi @cocciagialla, be careful when you update to the next beta release it'll be `@ApiSubresource`!

Also, I'm currently working on this  in #1245. I suggest that you subscribe to the issue to know when it'll be ready :).

cocciagialla commented:
@soyuka thanks, I am subscribed to #1245

jorgeluisacostaalonso commented:
There is a way that we can change the sub-queries made them in SubResourceDataProvider!

Right now it's possbile to change ou well add other conditions to main queryBuilder but the previousBuilder made them into SubResourceDataProvider we can't hook into them with any type of extension?
thanks
